### PR TITLE
ZCA Spectral Whitening of Input Features: decorrelate the 24-dim feature covariance

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current input preprocessing applies per-feature standardization: `x = (x - x_mean) / x_std`. This normalizes each of the 24 features independently but leaves **between-feature correlations intact**. The 24 input dimensions (xy coords, DSDF channels, TE frame distances, wake deficit, Re, gap, stagger, etc.) are highly correlated — all derived from a small number of underlying physical parameters (chord, AoA, Re, gap, stagger). This redundancy creates a poorly-conditioned optimization landscape.

**ZCA (Zero-phase Component Analysis) whitening** replaces per-feature standardization with a full covariance decorrelation: `x_whitened = (x - mu) @ W_zca.T`. The ZCA matrix `W_zca = U @ diag(1/sqrt(S+ε)) @ U.T` is precomputed once from training data and decorrelates ALL 24 feature dimensions simultaneously, producing an isotropic feature space where all directions are equally informative. This is the provably optimal preprocessing for gradient-based optimization (LeCun 1998, "Efficient BackProp").

**Why this fits our success lineage:** Our most consistent improvements come from data-representation changes: asinh pressure transform, TE coordinate frame, wake deficit feature. ZCA operates at the same level — transforming the input representation without any architectural change. It is **free at inference time** (a fixed linear transform).

**Why p_oodc:** ZCA ensures that OOD feature combinations (extreme Re, unseen gap values) are treated with equal importance in gradient space, rather than dominated by the largest-variance training-set directions.

**Key difference from per-foil whitening (PR #2261, in-flight):** That approach normalizes *output targets* per foil. This decorrelates *input features* globally. Complementary, not redundant.

**ZCA safety guarantee:** ZCA is an orthogonal rotation in feature space. The model can learn to invert it if ZCA is suboptimal — there is no regression floor.

## Instructions

Add ZCA whitening as an optional replacement for per-feature standardization. ~50 lines in `train.py`.

### Step 1: Add config flag

Near the other preprocessing flags, add:
```python
zca_whitening: bool = False
zca_eps: float = 1e-5  # regularization for near-zero singular values
```

### Step 2: Precompute ZCA matrix at training start

Find where `stats` is loaded (search for `x_mean` near dataset/stats setup). Add after stats loading, when `cfg.zca_whitening` is True:

```python
if cfg.zca_whitening:
    print("Computing ZCA whitening matrix from training data...")
    x_all = []
    for batch in train_loader:
        x_b = batch.x.float()
        x_all.append(x_b.reshape(-1, x_b.shape[-1]))
    X = torch.cat(x_all, dim=0)  # [N_total_nodes, 24]

    mu_zca = X.mean(0)  # [24]
    X_c = X - mu_zca
    C = (X_c.T @ X_c) / len(X_c)  # [24, 24]
    U, S, Vt = torch.linalg.svd(C)
    W_zca = U @ torch.diag(1.0 / (S + cfg.zca_eps).sqrt()) @ U.T  # [24, 24]
    zca_mu = mu_zca.to(device)
    zca_W = W_zca.to(device)
    print(f"ZCA condition number (before)={S[0]/S[-1]:.1f}, smallest_sv={S[-1]:.2e}")
```

**The ZCA pass over training data takes ~1-2 minutes (once). Budget this within the 180-min timeout.**

### Step 3: Apply ZCA in all forward passes

There are 4 occurrences of `x = (x - stats["x_mean"]) / stats["x_std"]` in train.py (training, validation, visualization variants). Replace each with:

```python
if cfg.zca_whitening:
    x = (x - zca_mu) @ zca_W.T
else:
    x = (x - stats["x_mean"]) / stats["x_std"]
```

**Note:** `zca_mu` replaces `stats["x_mean"]` as the centering term. Both are per-feature means — use `zca_mu` in the ZCA path for consistency.

### Step 4: Checkpoint

Save `zca_mu` and `zca_W` in the checkpoint:
```python
if cfg.zca_whitening:
    checkpoint['zca_mu'] = zca_mu.cpu()
    checkpoint['zca_W'] = zca_W.cpu()
```

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent fern --wandb_name "fern/zca-whitening-s42" \
  --wandb_group "round25/spectral-feature-whitening" \
  --seed 42 \
  --zca_whitening \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73: identical with --seed 73 --wandb_name "fern/zca-whitening-s73"
```

New flag vs baseline: `--zca_whitening` only.

### Step 6: If training is unstable (loss spikes in first 10 epochs)

Try `--lr 1.5e-4` — ZCA whitening changes the effective scale of the input, which can shift the optimal LR slightly. Report if you try this variant.

### Step 7: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds + 2-seed average vs baseline. Include W&B run IDs. Also report the ZCA condition number diagnostics from Step 2 output.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

Baseline W&B: `hgml7i2r` (seed 42), `qic03vrg` (seed 73) — PR #2213.

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent fern --wandb_name "fern/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### ZCA Spectral Whitening of Input Features

**ZCA Diagnostics (from debug run):**
- ZCA condition number (before whitening): ~7.9 billion
- Smallest singular value: ~3.25e-10
- This indicates near-singular input feature covariance — the 24 features have extreme collinearity

### Surface Pressure MAE (×10³ Pa)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Delta |
|--------|---------|---------|-----------|----------|-------|
| p_in   | 13.071  | 12.837  | **12.954** | 11.979  | +8.1% ❌ |
| p_oodc | 9.505   | 9.122   | **9.314**  | 7.643   | +21.9% ❌ |
| p_tan  | 31.321  | 30.737  | **31.029** | 28.341  | +9.5% ❌ |
| p_re   | 7.324   | 7.388   | **7.356**  | 6.300   | +16.8% ❌ |

### Volume MAE

| Metric | Seed 42 | Seed 73 | 2-seed avg |
|--------|---------|---------|-----------|
| vol_Ux (in) | 0.558 | 0.562 | 0.560 |
| vol_Uy (in) | 0.205 | 0.208 | 0.207 |
| vol_p (in)  | 13.175 | 13.138 | 13.157 |

### Training Details
- **Epochs:** 146 (s42), 147 (s73) — 180-min timeout
- **Memory:** 46.9 GB peak
- **W&B runs:** `p5lla3ly` (seed 42), `db3jcgz2` (seed 73)
- **W&B group:** `round25/spectral-feature-whitening`
- **New flag vs baseline:** `--zca_whitening` only

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/zca-whitening-s42" \
  --wandb_group "round25/spectral-feature-whitening" \
  --seed 42 --zca_whitening \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### What Happened

**ZCA whitening significantly degraded performance across all metrics.** The worst impact was on p_oodc (+22%), followed by p_re (+17%), p_tan (+10%), and p_in (+8%).

**Why it failed:**

1. **Near-singular covariance destroys information.** The input feature covariance has a condition number of ~7.9 billion. ZCA inverts this matrix (regularized with ε=1e-5), which amplifies noise in near-zero variance directions by factors of ~10⁵. The model receives heavily distorted features rather than a clean isotropic space.

2. **Per-feature standardization already handles scale.** The baseline's `(x - mean) / std` normalizes each feature independently. This is sufficient for features that are already well-scaled (coordinates, DSDF, Re, etc.). ZCA's full covariance decorrelation is overkill and introduces harmful amplification.

3. **OOD degradation is the worst** because ZCA is computed on training data. OOD samples (unseen conditions, Reynolds numbers) project poorly through a whitening matrix derived from the training distribution — the rotated feature space is actively misleading for these samples.

4. **Verify phase returned NaN for all metrics.** The verification path produced NaN pipeline MAE, likely because the ZCA transform interacts poorly with the verify code path's data loading (possibly missing nodes or different tensor shapes). This is a secondary bug that doesn't affect the training/validation metrics.

### Suggested Follow-ups

- **Do not pursue ZCA or PCA whitening** for this problem. The near-singular covariance makes full decorrelation harmful. Per-feature standardization is the right approach.
- **If decorrelation is desired**, a safer approach would be PCA with dimensionality reduction (dropping near-zero variance components) rather than full ZCA, but given the magnitude of degradation this is unlikely to help.
- **The verify NaN bug** should be investigated separately — it may indicate a code path that doesn't properly handle the ZCA tensors when no surface nodes exist in certain batches.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2266 because Closing — severe regression on all 4 metrics. ZCA whitening is harmful for this problem.

**2-seed avg vs current baseline (PR #2251):**

| Metric | Baseline | ZCA | Δ |
|--------|----------|-----|---|
| p_in | 11.891 | 12.954 | +8.9% ❌ |
| p_oodc | 7.561 | 9.314 | +23.2% ❌ |
| p_tan | 28.118 | 31.029 | +10.3% ❌ |
| p_re | 6.364 | 7.356 | +15.6% ❌ |

W&B: p5lla3ly (s42), db3jcgz2 (s73).

**Root cause:** The 24-dim input feature covariance has condition number ~7.9 billion — near-singular. ZCA inverts this (regularized with ε=1e-5), amplifying noise in near-zero variance directions by ~10⁵×. Per-feature standardization is the correct approach for this feature space. The OOD degradation is worst (+23% p_oodc) because the training-derived whitening matrix is maximally misleading for OOD feature combinations.

fern's analysis is spot-on. Adding ZCA and PCA whitening to the 'DO NOT REVISIT' list..
